core/src/main/java/org/elasticsearch/search/SearchModule.java

```diff
@@ -289,15 +294,14 @@ public void registerSuggester(String key, Suggester<?> suggester) {
     /**
      * Register a new ScoreFunctionParser.
      */
-    public void registerFunctionScoreParser(ScoreFunctionParser<? extends ScoreFunctionBuilder> parser) {
+    public <T extends ScoreFunctionBuilder<T>> void registerScoreFunction(ScoreFunctionParser<T> parser, Writeable.Reader<T> reader) {
         for (String name: parser.getNames()) {
-            Object oldValue = functionScoreParsers.putIfAbsent(name, parser);
+            Object oldValue = scoreFunctionParsers.putIfAbsent(name, parser);
             if (oldValue != null) {
                 throw new IllegalArgumentException("Function score parser [" + oldValue + "] already registered for name [" + name + "]");
             }
         }
-        @SuppressWarnings("unchecked") NamedWriteable<? extends ScoreFunctionBuilder> sfb = parser.getBuilderPrototype();
-        namedWriteableRegistry.registerPrototype(ScoreFunctionBuilder.class, sfb);
+        namedWriteableRegistry.register(ScoreFunctionBuilder.class, parser.getNames()[0], reader);
```

Comment:
I think this has the same problem as in #17458 as it uses the first parser name to register the named writeable.